### PR TITLE
contour: parameterise envoy scraping interval

### PR DESCRIPTION
**Pull request by surajssd**
This commit adds a parameter to contour's component config, using which now users can choose what (prometheus) scrape interval they want for envoy.

# How to use

Deploy contour using following config:

```tf
component "contour" {
  enable_monitoring = true

  envoy {
    metrics_scrape_interval = "30s"
  }
}
```

And you can watch the field `interval` vary according to the value of `metrics_scrape_interval`. If the field `metrics_scrape_interval` is not provided then the `interval` field is also left out.

```bash
kubectl -n projectcontour get servicemonitor envoy -o yaml | grep interval
```

Fixes: #1194

### CODE REVIEW

File: docs/configuration-reference/components/contour.md [hunk 1]
```diff
@@ -60,6 +64,7 @@ Table of all the arguments accepted by the component.
 | `node_affinity`     | Node affinity for deploying the operator pod and envoy daemonset.                                       |       -        | list(object({key = string, operator = string, values = list(string)}))                                         |  false   |
 | `service_type`      | The type of Kubernetes service used to expose Envoy. Set as "NodePort" on the **AWS** platform.         | "LoadBalancer" | string                                                                                                         |  false   |
 | `toleration`        | Tolerations that the operator and envoy pods will tolerate.                                             |       -        | list(object({key = string, effect = string, operator = string, value = string, toleration_seconds = string })) |  false   |
+| `envoy.metrics_scrape_interval` | Interval at which Prometheus will scrape Envoy. Valid only when `enable_monitoring` is set to `true`. |       -        | string                                                                                               |  false   |
```
Comment (invidian) [inline docs/configuration-reference/components/contour.md:67]: Table is not aligned + we should validate this document condition and have unit tests for it.

**Review by invidian — CHANGES_REQUESTED** [review]
I still have a feeling that:
- Increasing scrape interval might not be an optimal solution to solve the problem.
- Such customization seems very specific to me and having large amount of such will make it hard to maintain. I think we have to invest time in making generic customization mechanism. But even then, we will have to deprecate and eventually remove almost all of already added ones.

It would be good if someone experienced with Envoy and Prometheus could have a look at it.

**Comment by surajssd** [comment]
@invidian

> - Increasing scrape interval might not be an optimal solution to solve the problem.

This PR does not change anything unless user does. Any why reducing scrape interval might not solve the problem?

> - Such customization seems very specific to me and having large amount of such will make it hard to maintain. I think we have to invest time in making generic customization mechanism. But even then, we will have to deprecate and eventually remove almost all of already added ones.

What are you suggesting we do in this particular case? I agree we need a generic customization mechanism.

> It would be good if someone experienced with Envoy and Prometheus could have a look at it.

And who would that be?

**Comment by invidian** [comment]
>This PR does not change anything unless user does. Any why reducing scrape interval might not solve the problem?

I'm not saying it won't solve the problem. I'm saying that it might not be the optimal solution. In https://github.com/kinvolk/lokomotive/issues/1194#issuecomment-730302433 I suggested testing with different bucket sizes for latency metrics. Or maybe some new metric should be added to Envoy to allow spotting those latency spikes more easily.

Increasing scrape interval seems more like a brute-force workaround to me, rather than a proper solution to the problem.

>What are you suggesting we do in this particular case?

Merge if it's urgent enough, switch focus to generic customization mechanism if not urgent IMO.

**Review by iaguis — APPROVED** [review]
lgtm

**Pull request merged**